bug: crons with concurrency limits set cause the engine to crash

I have hatchet self-hosted in a K8s cluster.

Container Images:
- engine: `ghcr.io/hatchet-dev/hatchet/hatchet-engine:v0.26.1 `
- api: `ghcr.io/hatchet-dev/hatchet/hatchet-api:v0.26.1`
- rabbitmq: `docker.io/bitnami/rabbitmq:3.13.2-debian-12-r0`

SDK: Python - `hatchet-sdk-0.23.0` (`0.22.5` prior)

After version `0.23.0`, I've consistently run into the following issue when a cron task gets triggered, which then causes a reboot loop on the `engine` container:

```
2024-05-10T15:34:53.555Z INF workflow 491b44e5-34ad-4764-847b-fedf8f838362 has concurrency settings service=workflows-controller
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x78 pc=0x14053cf]

goroutine 41 [running]:
github.com/hatchet-dev/hatchet/internal/services/controllers/workflows.(*WorkflowsControllerImpl).scheduleGetGroupAction(0xc00372ef50, {0x1b11638?, 0xc000751320?}, 0x0)
	/hatchet/internal/services/controllers/workflows/queue.go:211 +0xcf
github.com/hatchet-dev/hatchet/internal/services/controllers/workflows.(*WorkflowsControllerImpl).handleWorkflowRunQueued(0xc00372ef50, {0x1b11590?, 0x3d2f720?}, 0xc000750900)
	/hatchet/internal/services/controllers/workflows/queue.go:72 +0x618
github.com/hatchet-dev/hatchet/internal/services/controllers/workflows.(*WorkflowsControllerImpl).handleTask(0xc0004b3ad0?, {0x1b11590, 0x3d2f720}, 0xc000750900)
	/hatchet/internal/services/controllers/workflows/controller.go:199 +0x117
github.com/hatchet-dev/hatchet/internal/services/controllers/workflows.(*WorkflowsControllerImpl).Start.func1(0xc0007ba000?)
	/hatchet/internal/services/controllers/workflows/controller.go:161 +0x90
github.com/hatchet-dev/hatchet/internal/msgqueue/rabbitmq.(*MessageQueueImpl).subscribe.func1.2({{0x1b0f940, 0xc00062c7e0}, 0x0, {0x0, 0x0}, {0x0, 0x0}, 0x0, 0x0, {0x0, ...}, ...})
	/hatchet/internal/msgqueue/rabbitmq/rabbitmq.go:502 +0x88b
created by github.com/hatchet-dev/hatchet/internal/msgqueue/rabbitmq.(*MessageQueueImpl).subscribe.func1 in goroutine 154
	/hatchet/internal/msgqueue/rabbitmq/rabbitmq.go:451 +0x5c6
```

I've attempted the following to debug:

- Delete all workflows, which deletes the cron schedules and allows the `engine` container to get back up.
- Recreate `rabbitmq`, including the persistent data, which doesn't do anything.

I am able to trigger the workflow manually, but whenever the cron schedule triggers the workflow, that issue occurs.

--- a/cmd/hatchet-engine/engine/run.go
+++ b/cmd/hatchet-engine/engine/run.go
@@ -141,6 +141,7 @@ func Run(ctx context.Context, cf *loader.ConfigLoader) error {
 
 	if sc.HasService("workflowscontroller") {
 		wc, err := workflows.New(
+			workflows.WithAlerter(sc.Alerter),
 			workflows.WithMessageQueue(sc.MessageQueue),
 			workflows.WithRepository(sc.EngineRepository),
 			workflows.WithLogger(sc.Logger),

--- a/internal/repository/prisma/get_group_key_run.go
+++ b/internal/repository/prisma/get_group_key_run.go
@@ -214,7 +214,7 @@ func (s *getGroupKeyRunRepository) GetGroupKeyRunForEngine(ctx context.Context, 
 	}
 
 	if len(res) == 0 {
-		return nil, nil
+		return nil, fmt.Errorf("could not find group key run %s", getGroupKeyRunId)
 	}
 
 	return res[0], nil

--- a/internal/repository/workflow_run.go
+++ b/internal/repository/workflow_run.go
@@ -77,6 +77,10 @@ func GetCreateWorkflowRunOptsFromManual(
 	input []byte,
 	additionalMetadata map[string]interface{},
 ) (*CreateWorkflowRunOpts, error) {
+	if input == nil {
+		input = []byte("{}")
+	}
+
 	opts := &CreateWorkflowRunOpts{
 		DisplayName:        StringPtr(getWorkflowRunDisplayName(workflowVersion.WorkflowName)),
 		WorkflowVersionId:  sqlchelpers.UUIDToStr(workflowVersion.WorkflowVersion.ID),
@@ -86,11 +90,9 @@ func GetCreateWorkflowRunOptsFromManual(
 		AdditionalMetadata: additionalMetadata,
 	}
 
-	if input != nil {
-		if workflowVersion.ConcurrencyLimitStrategy.Valid {
-			opts.GetGroupKeyRun = &CreateGroupKeyRunOpts{
-				Input: input,
-			}
+	if workflowVersion.ConcurrencyLimitStrategy.Valid {
+		opts.GetGroupKeyRun = &CreateGroupKeyRunOpts{
+			Input: input,
 		}
 	}
 
@@ -105,6 +107,10 @@ func GetCreateWorkflowRunOptsFromParent(
 	childKey *string,
 	additionalMetadata map[string]interface{},
 ) (*CreateWorkflowRunOpts, error) {
+	if input == nil {
+		input = []byte("{}")
+	}
+
 	opts := &CreateWorkflowRunOpts{
 		DisplayName:        StringPtr(getWorkflowRunDisplayName(workflowVersion.WorkflowName)),
 		WorkflowVersionId:  sqlchelpers.UUIDToStr(workflowVersion.WorkflowVersion.ID),
@@ -116,11 +122,9 @@ func GetCreateWorkflowRunOptsFromParent(
 
 	WithParent(parentId, parentStepRunId, childIndex, childKey)(opts)
 
-	if input != nil {
-		if workflowVersion.ConcurrencyLimitStrategy.Valid {
-			opts.GetGroupKeyRun = &CreateGroupKeyRunOpts{
-				Input: input,
-			}
+	if workflowVersion.ConcurrencyLimitStrategy.Valid {
+		opts.GetGroupKeyRun = &CreateGroupKeyRunOpts{
+			Input: input,
 		}
 	}
 
@@ -133,6 +137,10 @@ func GetCreateWorkflowRunOptsFromEvent(
 	input []byte,
 	additionalMetadata map[string]interface{},
 ) (*CreateWorkflowRunOpts, error) {
+	if input == nil {
+		input = []byte("{}")
+	}
+
 	opts := &CreateWorkflowRunOpts{
 		DisplayName:        StringPtr(getWorkflowRunDisplayName(workflowVersion.WorkflowName)),
 		WorkflowVersionId:  sqlchelpers.UUIDToStr(workflowVersion.WorkflowVersion.ID),
@@ -142,11 +150,9 @@ func GetCreateWorkflowRunOptsFromEvent(
 		AdditionalMetadata: additionalMetadata,
 	}
 
-	if input != nil {
-		if workflowVersion.ConcurrencyLimitStrategy.Valid {
-			opts.GetGroupKeyRun = &CreateGroupKeyRunOpts{
-				Input: input,
-			}
+	if workflowVersion.ConcurrencyLimitStrategy.Valid {
+		opts.GetGroupKeyRun = &CreateGroupKeyRunOpts{
+			Input: input,
 		}
 	}
 
@@ -160,6 +166,10 @@ func GetCreateWorkflowRunOptsFromCron(
 	input []byte,
 	additionalMetadata map[string]interface{},
 ) (*CreateWorkflowRunOpts, error) {
+	if input == nil {
+		input = []byte("{}")
+	}
+
 	opts := &CreateWorkflowRunOpts{
 		DisplayName:        StringPtr(getWorkflowRunDisplayName(workflowVersion.WorkflowName)),
 		WorkflowVersionId:  sqlchelpers.UUIDToStr(workflowVersion.WorkflowVersion.ID),
@@ -170,11 +180,9 @@ func GetCreateWorkflowRunOptsFromCron(
 		AdditionalMetadata: additionalMetadata,
 	}
 
-	if input != nil {
-		if workflowVersion.ConcurrencyLimitStrategy.Valid {
-			opts.GetGroupKeyRun = &CreateGroupKeyRunOpts{
-				Input: input,
-			}
+	if workflowVersion.ConcurrencyLimitStrategy.Valid {
+		opts.GetGroupKeyRun = &CreateGroupKeyRunOpts{
+			Input: input,
 		}
 	}
 
@@ -188,6 +196,10 @@ func GetCreateWorkflowRunOptsFromSchedule(
 	additionalMetadata map[string]interface{},
 	fs ...CreateWorkflowRunOpt,
 ) (*CreateWorkflowRunOpts, error) {
+	if input == nil {
+		input = []byte("{}")
+	}
+
 	opts := &CreateWorkflowRunOpts{
 		DisplayName:         StringPtr(getWorkflowRunDisplayName(workflowVersion.WorkflowName)),
 		WorkflowVersionId:   sqlchelpers.UUIDToStr(workflowVersion.WorkflowVersion.ID),
@@ -197,11 +209,9 @@ func GetCreateWorkflowRunOptsFromSchedule(
 		AdditionalMetadata:  additionalMetadata,
 	}
 
-	if input != nil {
-		if workflowVersion.ConcurrencyLimitStrategy.Valid {
-			opts.GetGroupKeyRun = &CreateGroupKeyRunOpts{
-				Input: input,
-			}
+	if workflowVersion.ConcurrencyLimitStrategy.Valid {
+		opts.GetGroupKeyRun = &CreateGroupKeyRunOpts{
+			Input: input,
 		}
 	}
 

--- a/internal/services/controllers/jobs/controller.go
+++ b/internal/services/controllers/jobs/controller.go
@@ -21,6 +21,7 @@ import (
 	"github.com/hatchet-dev/hatchet/internal/repository/prisma/dbsqlc"
 	"github.com/hatchet-dev/hatchet/internal/repository/prisma/sqlchelpers"
 	"github.com/hatchet-dev/hatchet/internal/services/shared/defaults"
+	"github.com/hatchet-dev/hatchet/internal/services/shared/recoveryutils"
 	"github.com/hatchet-dev/hatchet/internal/services/shared/tasktypes"
 	"github.com/hatchet-dev/hatchet/internal/telemetry"
 	"github.com/hatchet-dev/hatchet/internal/telemetry/servertel"
@@ -198,7 +199,17 @@ func (jc *JobsControllerImpl) Start() (func() error, error) {
 	return cleanup, nil
 }
 
-func (ec *JobsControllerImpl) handleTask(ctx context.Context, task *msgqueue.Message) error {
+func (ec *JobsControllerImpl) handleTask(ctx context.Context, task *msgqueue.Message) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			recoverErr := recoveryutils.RecoverWithAlert(ec.l, ec.a, r)
+
+			if recoverErr != nil {
+				err = recoverErr
+			}
+		}
+	}()
+
 	switch task.ID {
 	case "job-run-queued":
 		return ec.handleJobRunQueued(ctx, task)
@@ -1044,15 +1055,7 @@ func (ec *JobsControllerImpl) cancelStepRun(ctx context.Context, tenantId, stepR
 func (ec *JobsControllerImpl) handleStepRunUpdateInfo(stepRun *dbsqlc.GetStepRunForEngineRow, updateInfo *repository.StepRunUpdateInfo) {
 	defer func() {
 		if r := recover(); r != nil {
-			err, ok := r.(error)
-
-			if !ok {
-				err = fmt.Errorf("%v", r)
-			}
-
-			ec.l.Error().Err(err).Msg("recovered from panic")
-
-			return
+			recoveryutils.RecoverWithAlert(ec.l, ec.a, r) // nolint:errcheck
 		}
 	}()
 

--- a/internal/services/controllers/workflows/controller.go
+++ b/internal/services/controllers/workflows/controller.go
@@ -18,8 +18,10 @@ import (
 	"github.com/hatchet-dev/hatchet/internal/repository/prisma/db"
 	"github.com/hatchet-dev/hatchet/internal/repository/prisma/dbsqlc"
 	"github.com/hatchet-dev/hatchet/internal/repository/prisma/sqlchelpers"
+	"github.com/hatchet-dev/hatchet/internal/services/shared/recoveryutils"
 	"github.com/hatchet-dev/hatchet/internal/services/shared/tasktypes"
 	"github.com/hatchet-dev/hatchet/internal/telemetry"
+	hatcheterrors "github.com/hatchet-dev/hatchet/pkg/errors"
 )
 
 type WorkflowsController interface {
@@ -33,23 +35,28 @@ type WorkflowsControllerImpl struct {
 	dv            datautils.DataDecoderValidator
 	s             gocron.Scheduler
 	tenantAlerter *alerting.TenantAlertManager
+	a             *hatcheterrors.Wrapped
 }
 
 type WorkflowsControllerOpt func(*WorkflowsControllerOpts)
 
 type WorkflowsControllerOpts struct {
-	mq   msgqueue.MessageQueue
-	l    *zerolog.Logger
-	repo repository.EngineRepository
-	dv   datautils.DataDecoderValidator
-	ta   *alerting.TenantAlertManager
+	mq      msgqueue.MessageQueue
+	l       *zerolog.Logger
+	repo    repository.EngineRepository
+	dv      datautils.DataDecoderValidator
+	ta      *alerting.TenantAlertManager
+	alerter hatcheterrors.Alerter
 }
 
 func defaultWorkflowsControllerOpts() *WorkflowsControllerOpts {
 	logger := logger.NewDefaultLogger("workflows-controller")
+	alerter := hatcheterrors.NoOpAlerter{}
+
 	return &WorkflowsControllerOpts{
-		l:  &logger,
-		dv: datautils.NewDataDecoderValidator(),
+		l:       &logger,
+		dv:      datautils.NewDataDecoderValidator(),
+		alerter: alerter,
 	}
 }
 
@@ -68,6 +75,12 @@ func WithLogger(l *zerolog.Logger) WorkflowsControllerOpt {
 func WithRepository(r repository.EngineRepository) WorkflowsControllerOpt {
 	return func(opts *WorkflowsControllerOpts) {
 		opts.repo = r
+	}
+}
+
+func WithAlerter(a hatcheterrors.Alerter) WorkflowsControllerOpt {
+	return func(opts *WorkflowsControllerOpts) {
+		opts.alerter = a
 	}
 }
 
@@ -111,6 +124,9 @@ func New(fs ...WorkflowsControllerOpt) (*WorkflowsControllerImpl, error) {
 	newLogger := opts.l.With().Str("service", "workflows-controller").Logger()
 	opts.l = &newLogger
 
+	a := hatcheterrors.NewWrapped(opts.alerter)
+	a.WithData(map[string]interface{}{"service": "workflows-controller"})
+
 	return &WorkflowsControllerImpl{
 		mq:            opts.mq,
 		l:             opts.l,
@@ -118,6 +134,7 @@ func New(fs ...WorkflowsControllerOpt) (*WorkflowsControllerImpl, error) {
 		dv:            opts.dv,
 		s:             s,
 		tenantAlerter: opts.ta,
+		a:             a,
 	}, nil
 }
 
@@ -193,7 +210,17 @@ func (wc *WorkflowsControllerImpl) Start() (func() error, error) {
 	return cleanup, nil
 }
 
-func (wc *WorkflowsControllerImpl) handleTask(ctx context.Context, task *msgqueue.Message) error {
+func (wc *WorkflowsControllerImpl) handleTask(ctx context.Context, task *msgqueue.Message) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			recoverErr := recoveryutils.RecoverWithAlert(wc.l, wc.a, r)
+
+			if recoverErr != nil {
+				err = recoverErr
+			}
+		}
+	}()
+
 	switch task.ID {
 	case "workflow-run-queued":
 		return wc.handleWorkflowRunQueued(ctx, task)

--- a/internal/services/controllers/workflows/queue.go
+++ b/internal/services/controllers/workflows/queue.go
@@ -54,7 +54,7 @@ func (wc *WorkflowsControllerImpl) handleWorkflowRunQueued(ctx context.Context, 
 
 	// determine if we should start this workflow run or we need to limit its concurrency
 	// if the workflow has concurrency settings, then we need to check if we can start it
-	if workflowRun.ConcurrencyLimitStrategy.Valid {
+	if workflowRun.ConcurrencyLimitStrategy.Valid && workflowRun.GetGroupKeyRunId.Valid {
 		wc.l.Info().Msgf("workflow %s has concurrency settings", workflowRunId)
 
 		groupKeyRunId := sqlchelpers.UUIDToStr(workflowRun.GetGroupKeyRunId)
@@ -76,6 +76,8 @@ func (wc *WorkflowsControllerImpl) handleWorkflowRunQueued(ctx context.Context, 
 		}
 
 		return nil
+	} else if workflowRun.ConcurrencyLimitStrategy.Valid && !workflowRun.GetGroupKeyRunId.Valid {
+		return fmt.Errorf("workflow run %s has concurrency settings but no group key run", workflowRunId)
 	}
 
 	err = wc.queueWorkflowRunJobs(ctx, workflowRun)

--- a/internal/services/dispatcher/dispatcher.go
+++ b/internal/services/dispatcher/dispatcher.go
@@ -18,6 +18,7 @@ import (
 	"github.com/hatchet-dev/hatchet/internal/repository"
 	"github.com/hatchet-dev/hatchet/internal/repository/prisma/sqlchelpers"
 	"github.com/hatchet-dev/hatchet/internal/services/dispatcher/contracts"
+	"github.com/hatchet-dev/hatchet/internal/services/shared/recoveryutils"
 	"github.com/hatchet-dev/hatchet/internal/services/shared/tasktypes"
 	"github.com/hatchet-dev/hatchet/internal/telemetry"
 	"github.com/hatchet-dev/hatchet/internal/telemetry/servertel"
@@ -303,7 +304,11 @@ func (d *DispatcherImpl) Start() (func() error, error) {
 func (d *DispatcherImpl) handleTask(ctx context.Context, task *msgqueue.Message) (err error) {
 	defer func() {
 		if r := recover(); r != nil {
-			err = fmt.Errorf("recovered from panic: %v", r)
+			recoverErr := recoveryutils.RecoverWithAlert(d.l, d.a, r)
+
+			if recoverErr != nil {
+				err = recoverErr
+			}
 		}
 	}()
 

--- a/internal/services/shared/recoveryutils/recover.go
+++ b/internal/services/shared/recoveryutils/recover.go
@@ -1,0 +1,29 @@
+package recoveryutils
+
+import (
+	"fmt"
+	"runtime/debug"
+
+	"github.com/rs/zerolog"
+
+	hatcheterrors "github.com/hatchet-dev/hatchet/pkg/errors"
+)
+
+func RecoverWithAlert(l *zerolog.Logger, a *hatcheterrors.Wrapped, r any) error {
+	var ok bool
+	err, ok := r.(error)
+
+	if !ok {
+		err = fmt.Errorf("%v", r)
+	}
+
+	err = fmt.Errorf("recovered from panic: %w. Stack trace:\n%s", err, string(debug.Stack()))
+
+	l.Error().Err(err).Msgf("recovered from panic")
+
+	if a != nil {
+		return a.WrapErr(err, nil)
+	}
+
+	return err
+}

